lib/streamlit/elements/text_widgets.py

```diff
@@ -60,9 +82,28 @@ def serialize(self, v: str) -> str:
 class TextAreaSerde:
     value: SupportsStr
 
-    def deserialize(self, ui_value: Optional[str], widget_id: str = "") -> str:
+    def deserialize(self, ui_value: str | None, widget_id: str = "") -> str:
         return str(ui_value if ui_value is not None else self.value)
 
+    def deserialize_nullable(
```

Comment:
I'm a bit concerned about this solution here with the flag in `session_state`. The session state is more of a user-facing state; putting some internally used technical flags inside the `session_state` feels problematic. 

Wouldn't it be possible that we make the `value` field optional as well for the text_area / text_input message? And then in the frontend, if `clearable==True`, we always return the widget state with an unset value. It probably will also require changes to the [WidgetStates.proto](https://github.com/streamlit/streamlit/blob/develop/proto/streamlit/proto/WidgetStates.proto) and to the [widget state implementation](https://github.com/streamlit/streamlit/blob/33cd55cdda916d947aa3cfdbc5caee4bb80f2cdf/lib/streamlit/runtime/state/session_state.py#L89) in Python. But I think we can get a cleaner more unified solution implemented here on how widgets can communicate an empty state.

Reply:
@LukasMasuch agree its not a clean solution, the problem here is that we need to return `None` initially and `""` empty string afterwards.

The widget keeps the state of only one type and one variable, that how I understand current implementation.

We cannot keep it in proto since on every re-run it will behave the same and on first run we want it to return `None` and later on we want to return `""`. 

It's only the case with text widgets though, other do not require this kind of change, because they do not have this requirement
> if user typed sth before return "" instead of `None`

Do you think changing widget states additionally to all these changes will make this PR less complex? I'm moving to s4t project and I'd like to close it soon

Reply:
Yeah, removing this from the spec will make implementation much cleaner cc: @jrieke